File: static/js/robustness.js
/* ============================================================
 * robustness.js — 环境扰动模拟与稳健性评估模块
 * ============================================================ */
(function (global) {
  'use strict';

  const RobustnessApp = {};
  let ctx = null;

  const PARAM_LABELS = {
    temperature: '环境温度',
    viscosity: '液体黏度',
    inflow_amplitude: '注水波动幅度',
    orifice_wear: '孔径磨损程度',
    tilt_angle: '容器倾斜角度'
  };

  RobustnessApp.init = function (context) {
    ctx = context;
    ctx.robustnessConfig = null;
    ctx.robustnessAssessment = null;
    ctx.robustnessScenarios = [];
    ctx.robustnessIsMulti = false;
    RobustnessApp.bindEvents();
    RobustnessApp.loadConfig();
  };

  RobustnessApp.bindEvents = function () {
    const runBtn = document.getElementById('robustness-run-btn');
    if (runBtn) {
      runBtn.addEventListener('click', () => RobustnessApp.runSimulation());
    }
    const saveCfgBtn = document.getElementById('robustness-save-config-btn');
    if (saveCfgBtn) {
      saveCfgBtn.addEventListener('click', () => RobustnessApp.saveConfig());
    }
    const multiToggle = document.getElementById('robustness-multi-toggle');
    if (multiToggle) {
      multiToggle.addEventListener('change', (e) => {
        ctx.robustnessIsMulti = e.target.checked;
        RobustnessApp.loadAssessment();
      });
    }
    const scenarioSelect = document.getElementById('robustness-scenario-select');
    if (scenarioSelect) {
      scenarioSelect.addEventListener('change', (e) => {
        if (e.target.value) {
          RobustnessApp.loadScenarioDetail(parseInt(e.target.value));
        }
      });
    }
  };

  RobustnessApp.apiUrl = function (path) {
    return `/api/projects/${ctx.projectId}${path}`;
  };

  RobustnessApp.loadConfig = async function () {
    const { ok, data } = await __apiJson(RobustnessApp.apiUrl('/robustness/config'));
    if (ok && data?.ok) {
      ctx.robustnessConfig = data.config;
      RobustnessApp.renderConfigForm(data.config);
    }
    RobustnessApp.loadAssessment();
  };

  RobustnessApp.renderConfigForm = function (cfg) {
    const setVal = (id, val) => {
      const el = document.getElementById(id);
      if (el && val != null) el.value = val;
    };
    const setChecked = (id, val) => {
      const el = document.getElementById(id);
      if (el) el.checked = !!val;
    };
    setVal('rb-temperature-min', cfg.temperature_min);
    setVal('rb-temperature-max', cfg.temperature_max);
    setVal('rb-temperature-baseline', cfg.temperature_baseline);
    setChecked('rb-temperature-enabled', cfg.temperature_enabled);

    setVal('rb-viscosity-min', cfg.viscosity_min);
    setVal('rb-viscosity-max', cfg.viscosity_max);
    setVal('rb-viscosity-baseline', cfg.viscosity_baseline);
    setChecked('rb-viscosity-enabled', cfg.viscosity_enabled);

    setVal('rb-inflow-amplitude', cfg.inflow_fluctuation_amplitude);
    setVal('rb-inflow-frequency', cfg.inflow_fluctuation_frequency);
    setChecked('rb-inflow-enabled', cfg.inflow_fluctuation_enabled);

    setVal('rb-orifice-wear-rate', cfg.orifice_wear_rate);
    setVal('rb-orifice-wear-max', cfg.orifice_wear_max);
    setChecked('rb-orifice-enabled', cfg.orifice_wear_enabled);

    setVal('rb-tilt-min', cfg.tilt_angle_min);
    setVal('rb-tilt-max', cfg.tilt_angle_max);
    setVal('rb-tilt-baseline', cfg.tilt_angle_baseline);
    setChecked('rb-tilt-enabled', cfg.tilt_enabled);

    setVal('rb-duration', cfg.simulation_duration);
    setVal('rb-timestep', cfg.time_step);
    setVal('rb-scenario-count', cfg.scenario_count);
  };

  RobustnessApp.saveConfig = async function () {
    const getVal = (id) => {
      const el = document.getElementById(id);
      return el ? parseFloat(el.value) : null;
    };
    const getChecked = (id) => {
      const el = document.getElementById(id);
      return el ? el.checked : false;
    };
    const payload = {
      temperature_min: getVal('rb-temperature-min'),
      temperature_max: getVal('rb-temperature-max'),
      temperature_baseline: getVal('rb-temperature-baseline'),
      temperature_enabled: getChecked('rb-temperature-enabled'),
      viscosity_min: getVal('rb-viscosity-min'),
      viscosity_max: getVal('rb-viscosity-max'),
      viscosity_baseline: getVal('rb-viscosity-baseline'),
      viscosity_enabled: getChecked('rb-viscosity-enabled'),
      inflow_fluctuation_amplitude: getVal('rb-inflow-amplitude'),
      inflow_fluctuation_frequency: getVal('rb-inflow-frequency'),
      inflow_fluctuation_enabled: getChecked('rb-inflow-enabled'),
      orifice_wear_rate: getVal('rb-orifice-wear-rate'),
      orifice_wear_max: getVal('rb-orifice-wear-max'),
      orifice_wear_enabled: getChecked('rb-orifice-enabled'),
      tilt_angle_min: getVal('rb-tilt-min'),
      tilt_angle_max: getVal('rb-tilt-max'),
      tilt_angle_baseline: getVal('rb-tilt-baseline'),
      tilt_enabled: getChecked('rb-tilt-enabled'),
      simulation_duration: getVal('rb-duration'),
      time_step: getVal('rb-timestep'),
      scenario_count: getVal('rb-scenario-count'),
    };
    const btn = document.getElementById('robustness-save-config-btn');
    if (btn) { btn.disabled = true; btn.textContent = '保存中…'; }
    const { ok, data } = await __apiJson(RobustnessApp.apiUrl('/robustness/config'), {
      method: 'POST', body: payload
    });
    if (btn) { btn.disabled = false; btn.textContent = '保存扰动配置'; }
    if (ok && data?.ok) {
      ctx.robustnessConfig = data.config;
      __Toast.success('扰动配置已保存');
    } else {
      __Toast.error(data?.error || '保存失败');
    }
  };

  RobustnessApp.runSimulation = async function () {
    await RobustnessApp.saveConfig();
    const btn = document.getElementById('robustness-run-btn');
    if (btn) { btn.disabled = true; btn.textContent = '模拟中，请稍候…'; }
    const isMulti = ctx.robustnessIsMulti;
    const scenarioCount = ctx.robustnessConfig?.scenario_count || 50;
    const { ok, data } = await __apiJson(RobustnessApp.apiUrl('/robustness/simulate'), {
      method: 'POST', body: { is_multi_vessel: isMulti, scenario_count: scenarioCount }
    });
    if (btn) { btn.disabled = false; btn.textContent = '▶ 运行批量模拟'; }
    if (ok && data?.ok) {
      __Toast.success(`模拟完成：${data.result.completed}/${data.result.total_scenarios} 个场景`);
      await RobustnessApp.loadAssessment();
    } else {
      __Toast.error(data?.error || '模拟失败');
    }
  };

  RobustnessApp.loadScenarios = async function () {
    const isMulti = ctx.robustnessIsMulti;
    const { ok, data } = await __apiJson(
      RobustnessApp.apiUrl(`/robustness/scenarios?is_multi_vessel=${isMulti}`)
    );
    if (ok && data?.ok) {
      ctx.robustnessScenarios = data.scenarios;
      RobustnessApp.renderScenarioList(data.scenarios);
    }
  };

  RobustnessApp.renderScenarioList = function (scenarios) {
    const sel = document.getElementById('robustness-scenario-select');
    if (!sel) return;
    sel.innerHTML = '<option value="">-- 选择场景查看详情 --</option>' +
      scenarios.map(s => `<option value="${s.id}">${s.name}</option>`).join('');
  };

  RobustnessApp.loadScenarioDetail = async function (scenarioId) {
    const { ok, data } = await __apiJson(`/api/robustness/scenarios/${scenarioId}`);
    if (ok && data?.ok) {
      RobustnessApp.renderScenarioDetail(data.detail);
    }
  };

  RobustnessApp.renderScenarioDetail = function (detail) {
    const scenario = detail.scenario;
    const container = document.getElementById('robustness-scenario-detail');
    if (!container) return;
    let html = `<div class="scenario-meta"><strong>${scenario.name}</strong>`;
    const parts = [];
    if (scenario.temperature != null) parts.push(`温度 ${scenario.temperature}°C`);
    if (scenario.viscosity != null) parts.push(`黏度 x${scenario.viscosity}`);
    if (scenario.inflow_amplitude != null) parts.push(`注水波动 ±${(scenario.inflow_amplitude*100).toFixed(0)}%`);
    if (scenario.orifice_wear != null) parts.push(`孔径磨损 ${(scenario.orifice_wear*100).toFixed(0)}%`);
    if (scenario.tilt_angle != null) parts.push(`倾斜 ${scenario.tilt_angle}°`);
    html += ` <span class="muted">${parts.join(' · ')}</span></div>`;
    container.innerHTML = html;

    if (detail.vessel_results) {
      const traces = [];
      const names = Object.keys(detail.vessel_results);
      names.forEach((name, idx) => {
        const pts = detail.vessel_results[name];
        const color = ['#5C4033','#4A7C59','#C23B22','#B8860B','#4E6BA0'][idx % 5];
        traces.push({
          x: pts.map(p => p.time_point),
          y: pts.map(p => p.water_level),
          type: 'scatter', mode: 'lines', name: `${name} 水位`,
          line: { color, width: 2 }
        });
        if (pts.some(p => p.expected_level != null)) {
          traces.push({
            x: pts.map(p => p.time_point),
            y: pts.map(p => p.expected_level),
            type: 'scatter', mode: 'lines', name: `${name} 理论值`,
            line: { color, width: 1, dash: 'dash' }
          });
        }
      });
      const layout = __makeChartLayout('各容器水位曲线', '时间(分)', '水位(ml)', 340);
      Plotly.newPlot('robustness-scenario-chart', traces, layout, { displayModeBar: false, responsive: true });
    } else if (detail.results && detail.results.length) {
      const pts = detail.results;
      const traces = [{
        x: pts.map(p => p.time_point),
        y: pts.map(p => p.water_level),
        type: 'scatter', mode: 'lines', name: '实际水位',
        line: { color: '#5C4033', width: 2 }
      }];
      if (pts.some(p => p.expected_level != null)) {
        traces.push({
          x: pts.map(p => p.time_point),
          y: pts.map(p => p.expected_level),
          type: 'scatter', mode: 'lines', name: '理论水位',
          line: { color: '#4A7C59', width: 1, dash: 'dash' }
        });
      }
      const layout = __makeChartLayout('单壶水位变化曲线', '时间(分)', '水位(ml)', 340);
      Plotly.newPlot('robustness-scenario-chart', traces, layout, { displayModeBar: false, responsive: true });
    }
  };

  RobustnessApp.loadAssessment = async function () {
    const isMulti = ctx.robustnessIsMulti;
    await RobustnessApp.loadScenarios();
    const { ok, data } = await __apiJson(
      RobustnessApp.apiUrl(`/robustness/assessment?is_multi_vessel=${isMulti}`)
    );
    if (ok && data?.ok) {
      ctx.robustnessAssessment = data.assessment;
      RobustnessApp.renderAssessment(data.assessment);
    } else {
      RobustnessApp.renderEmptyAssessment();
    }
  };

  RobustnessApp.renderEmptyAssessment = function () {
    const scoreEl = document.getElementById('rb-score');
    if (scoreEl) scoreEl.textContent = '--';
    const gradeEl = document.getElementById('rb-grade');
    if (gradeEl) gradeEl.textContent = '未评估';
    ['rb-avg-err','rb-max-err','rb-err-std','rb-fail-rate'].forEach(id => {
      const el = document.getElementById(id);
      if (el) el.textContent = '--';
    });
    const summary = document.getElementById('rb-summary');
    if (summary) summary.innerHTML = '<div class="placeholder-cell">请先点击"运行批量模拟"生成评估结果</div>';
    const sensContainer = document.getElementById('rb-sensitivity');
    if (sensContainer) sensContainer.innerHTML = '<div class="placeholder-cell">暂无敏感度数据</div>';
    const rankingContainer = document.getElementById('rb-ranking');
    if (rankingContainer) rankingContainer.innerHTML = '<div class="placeholder-cell">暂无参数排序</div>';
    const adviceContainer = document.getElementById('rb-advice');
    if (adviceContainer) adviceContainer.innerHTML = '<div class="placeholder-cell">暂无校准建议</div>';
    RobustnessApp.renderScenariosChart([]);
  };

  RobustnessApp.renderAssessment = function (assessment) {
    const score = assessment.overall_stability_score || 0;
    const scoreEl = document.getElementById('rb-score');
    if (scoreEl) {
      scoreEl.textContent = score.toFixed(1);
      scoreEl.style.color = score >= 70 ? 'var(--c-accent)' : (score >= 50 ? 'var(--c-warn)' : 'var(--c-danger)');
    }
    const gradeEl = document.getElementById('rb-grade');
    if (gradeEl) {
      let grade = '较差';
      if (score >= 85) grade = '优秀';
      else if (score >= 70) grade = '良好';
      else if (score >= 50) grade = '一般';
      gradeEl.textContent = grade;
    }
    const setText = (id, val, suffix = '') => {
      const el = document.getElementById(id);
      if (el) el.textContent = val != null ? val.toFixed(2) + suffix : '--';
    };
    setText('rb-avg-err', assessment.avg_error, '%');
    setText('rb-max-err', assessment.max_error, '%');
    setText('rb-err-std', assessment.error_std, '%');
    const failEl = document.getElementById('rb-fail-rate');
    if (failEl) failEl.textContent = assessment.failure_rate != null ? (assessment.failure_rate * 100).toFixed(1) + '%' : '--';

    const summary = document.getElementById('rb-summary');
    if (summary) {
      summary.innerHTML = `<div class="assessment-summary">${assessment.summary || ''}</div>`;
    }

    RobustnessApp.renderSensitivityChart(assessment.sensitivity_scores || []);
    RobustnessApp.renderParameterRanking(assessment.parameter_ranking || []);
    RobustnessApp.renderCalibrationAdvice(assessment.calibration_advice || []);
    RobustnessApp.renderScenariosChart(assessment.scenario_summaries || []);
  };

  RobustnessApp.renderSensitivityChart = function (scores) {
    const container = document.getElementById('rb-sensitivity');
    if (!container) return;
    if (!scores.length) {
      container.innerHTML = '<div class="placeholder-cell">暂无敏感度数据</div>';
      return;
    }
    const labels = scores.map(s => s.parameter_label);
    const values = scores.map(s => s.score);
    const colors = scores.map(s => s.score >= 50 ? 'rgba(194,59,34,0.8)' : (s.score >= 25 ? 'rgba(180,112,26,0.8)' : 'rgba(74,124,89,0.8)'));
    const trace = {
      x: values,
      y: labels,
      type: 'bar',
      orientation: 'h',
      marker: { color: colors },
      text: values.map(v => v.toFixed(1)),
      textposition: 'auto',
    };
    const layout = __makeChartLayout('参数敏感度评分', '敏感度得分', '', 240 + scores.length * 28);
    layout.xaxis.range = [0, 100];
    Plotly.newPlot('rb-sensitivity', [trace], layout, { displayModeBar: false, responsive: true });
  };

  RobustnessApp.renderParameterRanking = function (ranking) {
    const container = document.getElementById('rb-ranking');
    if (!container) return;
    if (!ranking.length) {
      container.innerHTML = '<div class="placeholder-cell">暂无参数排序</div>';
      return;
    }
    let html = '<table class="interval-error-table"><thead><tr><th>排名</th><th>参数</th><th>类别</th><th>敏感度</th><th>说明</th></tr></thead><tbody>';
    ranking.forEach(r => {
      html += `<tr>
        <td><strong>#${r.rank}</strong></td>
        <td>${r.parameter_label}</td>
        <td><span class="badge badge-subtle">${r.category}</span></td>
        <td>${r.sensitivity_score.toFixed(1)}</td>
        <td class="muted">${r.description}</td>
      </tr>`;
    });
    html += '</tbody></table>';
    container.innerHTML = html;
  };

  RobustnessApp.renderCalibrationAdvice = function (advice) {
    const container = document.getElementById('rb-advice');
    if (!container) return;
    if (!advice.length) {
      container.innerHTML = '<div class="placeholder-cell">系统对当前扰动范围不敏感，无需特殊校准措施</div>';
      return;
    }
    let html = '';
    advice.forEach(a => {
      const pClass = a.priority === 'high' ? 'danger' : (a.priority === 'medium' ? 'warn' : 'subtle');
      const pLabel = a.priority === 'high' ? '高优先级' : (a.priority === 'medium' ? '中优先级' : '低优先级');
      html += `<div class="advice-card priority-${a.priority}">
        <div class="advice-header">
          <span class="badge badge-${pClass}">${pLabel}</span>
          <strong>${a.parameter_label}</strong>
          <span class="muted">预期改善 ${a.expected_improvement.toFixed(1)}%</span>
        </div>
        <div class="advice-action">${a.action}</div>
        <div class="advice-target">目标范围：${a.target_range}</div>
        <div class="advice-reason muted">${a.reason}</div>
      </div>`;
    });
    container.innerHTML = html;
  };

  RobustnessApp.renderScenariosChart = function (summaries) {
    const container = document.getElementById('rb-scenarios-chart');
    if (!container) return;
    if (!summaries.length) {
      container.innerHTML = '<div class="placeholder-cell">暂无场景数据</div>';
      return;
    }
    const sorted = [...summaries].sort((a, b) => a.avg_error - b.avg_error);
    const labels = sorted.map(s => s.name.length > 20 ? s.name.substring(0, 20) + '…' : s.name);
    const avgErrors = sorted.map(s => s.avg_error);
    const maxErrors = sorted.map(s => s.max_error);
    const colors = sorted.map(s => s.failed ? 'rgba(194,59,34,0.8)' : 'rgba(74,124,89,0.6)');
    const traces = [
      {
        x: labels, y: avgErrors, type: 'bar', name: '平均误差(%)',
        marker: { color: colors }
      },
      {
        x: labels, y: maxErrors, type: 'scatter', mode: 'markers',
        name: '最大误差(%)', marker: { color: '#C23B22', size: 8 }
      }
    ];
    const layout = __makeChartLayout('各场景误差对比', '场景', '误差(%)', 360);
    layout.xaxis.tickangle = -30;
    Plotly.newPlot('rb-scenarios-chart', traces, layout, { displayModeBar: false, responsive: true });
  };

  global.RobustnessApp = RobustnessApp;
})(window);

File: schemas/robustness.py
from __future__ import annotations

from datetime import datetime
from typing import Optional, List, Dict, Any

from pydantic import BaseModel, Field, ConfigDict, model_validator


class PerturbationConfigUpdate(BaseModel):
    temperature_min: Optional[float] = Field(default=10.0, ge=-20, le=80)
    temperature_max: Optional[float] = Field(default=40.0, ge=-20, le=80)
    temperature_baseline: Optional[float] = Field(default=20.0, ge=-20, le=80)
    temperature_enabled: Optional[bool] = True

    viscosity_min: Optional[float] = Field(default=0.8, ge=0.1, le=10.0)
    viscosity_max: Optional[float] = Field(default=1.8, ge=0.1, le=10.0)
    viscosity_baseline: Optional[float] = Field(default=1.0, ge=0.1, le=10.0)
    viscosity_enabled: Optional[bool] = True

    inflow_fluctuation_amplitude: Optional[float] = Field(default=0.1, ge=0.0, le=1.0)
    inflow_fluctuation_frequency: Optional[float] = Field(default=0.5, ge=0.0, le=10.0)
    inflow_fluctuation_enabled: Optional[bool] = True

    orifice_wear_rate: Optional[float] = Field(default=0.02, ge=0.0, le=1.0)
    orifice_wear_max: Optional[float] = Field(default=0.3, ge=0.0, le=1.0)
    orifice_wear_enabled: Optional[bool] = True

    tilt_angle_min: Optional[float] = Field(default=0.0, ge=0.0, le=45.0)
    tilt_angle_max: Optional[float] = Field(default=5.0, ge=0.0, le=45.0)
    tilt_angle_baseline: Optional[float] = Field(default=0.0, ge=0.0, le=45.0)
    tilt_enabled: Optional[bool] = True

    simulation_duration: Optional[float] = Field(default=60.0, ge=1.0, le=1440.0)
    time_step: Optional[float] = Field(default=0.5, ge=0.01, le=10.0)
    scenario_count: Optional[int] = Field(default=50, ge=1, le=500)

    @model_validator(mode="after")
    def check_ranges(self) -> "PerturbationConfigUpdate":
        errors = []
        if self.temperature_min is not None and self.temperature_max is not None:
            if self.temperature_min > self.temperature_max:
                errors.append(f"温度最小值 ({self.temperature_min}) 不能大于最大值 ({self.temperature_max})")
        if self.temperature_baseline is not None:
            if self.temperature_min is not None and self.temperature_baseline < self.temperature_min:
                errors.append(f"温度基准值 ({self.temperature_baseline}) 不能小于最小值 ({self.temperature_min})")
            if self.temperature_max is not None and self.temperature_baseline > self.temperature_max:
                errors.append(f"温度基准值 ({self.temperature_baseline}) 不能大于最大值 ({self.temperature_max})")
        if self.viscosity_min is not None and self.viscosity_max is not None:
            if self.viscosity_min > self.viscosity_max:
                errors.append(f"黏度最小值 ({self.viscosity_min}) 不能大于最大值 ({self.viscosity_max})")
        if self.viscosity_baseline is not None:
            if self.viscosity_min is not None and self.viscosity_baseline < self.viscosity_min:
                errors.append(f"黏度基准值 ({self.viscosity_baseline}) 不能小于最小值 ({self.viscosity_min})")
            if self.viscosity_max is not None and self.viscosity_baseline > self.viscosity_max:
                errors.append(f"黏度基准值 ({self.viscosity_baseline}) 不能大于最大值 ({self.viscosity_max})")
        if self.tilt_angle_min is not None and self.tilt_angle_max is not None:
            if self.tilt_angle_min > self.tilt_angle_max:
                errors.append(f"倾斜角最小值 ({self.tilt_angle_min}) 不能大于最大值 ({self.tilt_angle_max})")
        if self.tilt_angle_baseline is not None:
            if self.tilt_angle_min is not None and self.tilt_angle_baseline < self.tilt_angle_min:
                errors.append(f"倾斜角基准值 ({self.tilt_angle_baseline}) 不能小于最小值 ({self.tilt_angle_min})")
            if self.tilt_angle_max is not None and self.tilt_angle_baseline > self.tilt_angle_max:
                errors.append(f"倾斜角基准值 ({self.tilt_angle_baseline}) 不能大于最大值 ({self.tilt_angle_max})")
        if self.simulation_duration is not None and self.time_step is not None:
            if self.time_step > self.simulation_duration:
                errors.append(f"时间步长 ({self.time_step}) 不能大于模拟时长 ({self.simulation_duration})")
        if errors:
            raise ValueError("; ".join(errors))
        return self


class PerturbationConfigOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    temperature_min: float
    temperature_max: float
    temperature_baseline: float
    temperature_enabled: bool

    viscosity_min: float
    viscosity_max: float
    viscosity_baseline: float
    viscosity_enabled: bool

    inflow_fluctuation_amplitude: float
    inflow_fluctuation_frequency: float
    inflow_fluctuation_enabled: bool

    orifice_wear_rate: float
    orifice_wear_max: float
    orifice_wear_enabled: bool

    tilt_angle_min: float
    tilt_angle_max: float
    tilt_angle_baseline: float
    tilt_enabled: bool

    simulation_duration: float
    time_step: float
    scenario_count: int


class SimulationScenarioOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    config_id: Optional[int] = None
    scenario_index: int
    name: str
    is_multi_vessel: bool
    status: str
    created_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    temperature: Optional[float] = None
    viscosity: Optional[float] = None
    inflow_amplitude: Optional[float] = None
    orifice_wear: Optional[float] = None
    tilt_angle: Optional[float] = None
    params: Optional[Dict[str, Any]] = None


class SimulationResultPoint(BaseModel):
    time_point: float
    water_level: float
    expected_level: Optional[float] = None
    flow_rate: Optional[float] = None
    inflow_rate: Optional[float] = None
    time_error: Optional[float] = None
    level_error: Optional[float] = None


class SimulationScenarioDetail(BaseModel):
    scenario: SimulationScenarioOut
    results: List[SimulationResultPoint] = []
    vessel_results: Optional[Dict[str, List[SimulationResultPoint]]] = None


class SensitivityScore(BaseModel):
    parameter: str
    parameter_label: str
    score: float
    correlation: float
    p_value: Optional[float] = None
    impact_direction: str


class ParameterRanking(BaseModel):
    rank: int
    parameter: str
    parameter_label: str
    sensitivity_score: float
    category: str
    description: str


class CalibrationAdvice(BaseModel):
    parameter: str
    parameter_label: str
    priority: str
    action: str
    target_range: str
    expected_improvement: float
    reason: str


class ScenarioSummary(BaseModel):
    scenario_id: int
    scenario_index: int
    name: str
    avg_error: float
    max_error: float
    final_level_error: float
    failed: bool
    params: Dict[str, Any]


class RobustnessAssessmentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    config_id: Optional[int] = None
    is_multi_vessel: bool
    created_at: Optional[datetime] = None

    overall_stability_score: Optional[float] = None
    avg_error: Optional[float] = None
    max_error: Optional[float] = None
    error_std: Optional[float] = None
    failure_rate: Optional[float] = None

    sensitivity_scores: Optional[List[SensitivityScore]] = None
    parameter_ranking: Optional[List[ParameterRanking]] = None
    calibration_advice: Optional[List[CalibrationAdvice]] = None
    scenario_summaries: Optional[List[ScenarioSummary]] = None
    summary: Optional[str] = None


class SimulationRunRequest(BaseModel):
    is_multi_vessel: bool = False
    scenario_count: Optional[int] = None


class BatchSimulationOut(BaseModel):
    ok: bool
    total_scenarios: int
    completed: int
    assessment_id: Optional[int] = None
    message: str
